lib/plugins/aws/provider.js

```diff
@@ -1700,4 +1740,232 @@ class AwsProvider {
   }
 }
 
+Object.defineProperties(
+  AwsProvider.prototype,
+  memoizeeMethods({
+    getAccountId: d(
+      async function () {
+        const result = await this.getAccountInfo();
+        return result.accountId;
+      },
+      { promise: true }
+    ),
+    ensureDockerIsAvailable: d(
+      async function () {
+        try {
+          await spawnExt('docker', ['--version']);
+        } catch (err) {
+          throw new this.serverless.classes.Error(
+            'Could not find Docker installation. Ensure Docker is installed before continuing.',
+            'DOCKER_COMMAND_NOT_AVAILABLE'
+          );
+        }
+      },
+      { promise: true }
+    ),
+    dockerLoginToEcr: d(
+      async function () {
+        const registryId = await this.getAccountId();
+        const result = await this.request('ECR', 'getAuthorizationToken', {
+          registryIds: [registryId],
+        });
+        const { authorizationToken, proxyEndpoint } = result.authorizationData[0];
+        const decodedAuthToken = Buffer.from(authorizationToken, 'base64').toString().split(':')[1];
+        const dockerArgs = [
+          'login',
+          '--username',
+          'AWS',
+          '--password',
+          decodedAuthToken,
+          proxyEndpoint,
+        ];
+        try {
+          const { stdBuffer } = await spawnExt('docker', dockerArgs);
+          process.stdout.write(stdBuffer);
```

Comment:
Is the warning that we use `--password` now visible in a console?

If it's the case I would probably not show output of this command by default (what's more important is image operations actions, where it's simple login)

Reply:
Yes, it's visible in the console, we could hide it but I'm not sure if we should hide information that we're storing the credentials in an unencrypted way locally. That's what is showing right now from `docker login`
```
WARNING! Using --password via the CLI is insecure. Use --password-stdin.
WARNING! Your password will be stored unencrypted in /home/pgrzesik/.docker/config.json.
Configure a credential helper to remove this warning. See
https://docs.docker.com/engine/reference/commandline/login/#credentials-store

Login Succeeded
```

Reply:
Hmm.. I think it's confusing for user. It's not user that made a choice of how password is transferred, and visible warning may imply to him that something is not right.

I would hide this output and instead show a visible notice:

```
Logged Docker successfully into AWS ECR.

Note: that authentication token is stored as unencrypted  in /home/pgrzesik/.docker/config.json.
```

Second line I'd put in orange as we do with other warnings.  I believe we should also add a note about that into doc.

Also this triggers another question, if user will configure a credential helper that's mentioned in docs, can we somehow omit the login step? (we may ofc leave it for other PR)

Reply:
About displaying that note: it might be tricky as the outputs vary between operating systems (as well as the location of `.docker/config.json`). As for the credential helpers, I believe we still have to perform login for two reasons:
1) Tokens obtained from AWS are short-lived and there's no information about their validity stored (we could try to fallback to login only if needed (when push operation is denied e.g.), but that still requires logging in)  
2) Even if credential helper is used (afaik), it's only about where the tokens are stored, e.g. in some kind of encrypted file instead of plaintext as in default, we still have to `login`, but it's just up to `docker` to properly store that token during `login`

Reply:
Now I'm thinking about it more - maybe we shouldn't start with `docker login`, but rather only issue it if `docker push` command fails due to incorrect credentials?

Reply:
There's also https://github.com/awslabs/amazon-ecr-credential-helper utility that would make `docker login` step redundant in case someone uses it

Reply:
> but rather only issue it if docker push command fails due to incorrect credentials?

Good idea! Going that road will definitely be nicer. At least user will have a choice and be able to setup authentication via own means if that's preferred.

but if for some reason recognition of authentication error is tricky I would skip it for now.

> About displaying that note: it might be tricky as the outputs vary between operating systems

I would probably then search for a phrase `stored unencrypted in <somePath>`, and show our note only if we find it in output, otherwise not show it at all.

Reply:
I've went for this type of output (warning with orange colour):

```
Serverless: Login to Docker succeeded!
Serverless: WARNING: Docker authentication token will be stored unencrypted. Configure Docker credential helper to remove this warning.
The push refers to repository [600238737408.dkr.ecr.us-east-1.amazonaws.com/serverless-someproject-dev]
bdfa6eb032c6: Preparing
61b3904fa195: Preparing
6a7f9b3693e2: Preparing
```

What do you think?

Reply:
Looks great, I would probably add `..will be stored unencrypted in docker config`, so user is not confused where it's stored in general

Reply:
good call :+1: